File: src/year2024/day9/python/part1.py
from collections import defaultdict

# line = open("../test1.txt").read().strip()
line = open("../prod1.txt").read().strip()

ans = 0
res = []
curr  = 0

for i in range(1, len(line), 2):
    for _ in range(int(line[i-1])):
        res.append(str(curr))

    for _ in range(int(line[i])):
        res.append(".")
    curr += 1

if len(line) % 2 != 0:
    for _ in range(int(line[-1])):
        res.append(str(curr))

l = 0
r = len(res) - 1
while l < r:
    if res[l] == ".":
        res[l], res[r] = res[r], "."
        r -= 1
    else:
        l += 1

for idx, n in enumerate(res):
    if n ==".":
        continue
    ans += (int(n) * idx)

print(ans)

File: src/year2024/day1/go/part1.go

package main

import (
    "fmt"
    "strings"
    "sort"
    "strconv"
    "os"
    "bufio"
)

func main() {
    file, _ := os.Open("../prod1.txt")
    defer file.Close()

    var lines []string
    scanner := bufio.NewScanner(file)
    for scanner.Scan() {
        lines = append(lines, scanner.Text())
    }

    var l, r []int

    for _, line := range lines {
        nums := strings.Fields(line)
        if value, err := strconv.Atoi(nums[0]); err == nil {
            l = append(l, value)

        }

        if value, err := strconv.Atoi(nums[1]); err == nil {
            r = append(r, value)
        }
    }

    // Sorts in place
    sort.Ints(l)
    sort.Ints(r)

    var ans int = 0

    for i := 0; i < len(l); i++ {
        diff := l[i] - r[i]
        ans += max(diff, -diff)

    }

    fmt.Println("Ans: ", ans)
}


